### PR TITLE
Refine anonymizer processing logics

### PR DESCRIPTION
**Pull request by BoyaWu10**
Refine anonymizer processing logics to enhance compatibility and performance:
- Remove unnecessary casts between `ITypedElement` and `ElementNode`
- Wrap different exceptions for invalid inputs to `InvalidInputException`
- Given JSON string input, parse it to `ITypedElement` directly instead of parsing it to POCO first to enhance performance
- Add cache for `nodesByType` and `nodesByName` to enhance performance

### CODE REVIEW

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Processors/SubstituteProcessor.cs [hunk 1]
```diff
@@ -52,12 +50,12 @@ public ProcessResult Process(ElementNode node, ProcessContext context = null, Di
             // Retrieve all nodes that have been processed before to keep 
             _ = GenerateKeepNodeSetForSubstitution(node, context.VisitedNodes, keepNodes);
             var processResult = SubstituteNode(node, replacementNode, context.VisitedNodes, keepNodes);
-            MarkSubstitutedFragmentAsVisited(node, context.VisitedNodes);
+            context.VisitedNodes.UnionWith(node.Descendants().CastElementNodes());
```
Comment (moria97) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Processors/SubstituteProcessor.cs:53]: 👍

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.R4.FunctionalTests/VersionSpecificTests.cs [hunk 2]
```diff
@@ -84,7 +84,7 @@ public void GivenCommonResourceWithStu3OnlyValue_WhenAnonymizing_ExceptionShould
             AnonymizerEngine engine = new AnonymizerEngine("r4-configuration-sample.json");
             string testContent = File.ReadAllText(ResourceTestsFile(testFile));
             
-            Assert.Throws<FormatException>(() => engine.AnonymizeJson(testContent));
+            Assert.Throws<StructuralTypeException>(() => engine.AnonymizeJson(testContent));
```
Comment (moria97) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.R4.FunctionalTests/VersionSpecificTests.cs:87]: Do we need to wrap the ```StructuralException ``` as some ```InvalidInputException```?

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs [hunk 3]
```diff
@@ -85,34 +89,128 @@ private ProcessResult ProcessResourceNode(ElementNode node)
...
+        }
 
-            return result;
+        private IEnumerable<ITypedElement> GetMatchNodes(AnonymizationFhirPathRule rule, ITypedElement node)
+        {
+            var matchNodes = new List<ITypedElement>();
+            var resourceTypeMatch = AnonymizationFhirPathRule.TypeRuleRegex.Match(rule.Path);
+            var nameMatch = AnonymizationFhirPathRule.NameRuleRegex.Match(rule.Path);
+
+            if (resourceTypeMatch.Success)
```
Comment (moria97) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs:152]: Can we reuse the logic in matching by type and name here?

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerConfigurations/AnonymizationFhirPathRule.cs [hunk 4]
```diff
@@ -6,19 +6,25 @@ namespace Microsoft.Health.Fhir.Anonymizer.Core.AnonymizerConfigurations
 {
     public class AnonymizationFhirPathRule : AnonymizerRule
     {
+        // Example: nodesByType('Patient').use
+        public static readonly Regex TypeRuleRegex = new Regex(@"^nodesByType\([\'|\""](?<type>[A-Za-z0-9]+)[\'|\""]\)(\.(?<expression>[A-Za-z0-9]+))?$");
```
Comment (moria97) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerConfigurations/AnonymizationFhirPathRule.cs:10]: Do we support FHIRPath functions here?

Comment (BoyaWu10) [reply FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerConfigurations/AnonymizationFhirPathRule.cs:10]: No, only children expression is supported. FHIRPath functions like `nodesByType("HumanName").nodesByType("Period")` will not work here.

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerConfigurations/AnonymizationFhirPathRule.cs [hunk 5]
```diff
@@ -6,19 +6,25 @@ namespace Microsoft.Health.Fhir.Anonymizer.Core.AnonymizerConfigurations
...
 
-        public bool IsResourceTypeRule { get { return Path.Equals(ResourceType);  } }
+        public bool IsResourceTypeRule => Path.Equals(ResourceType);
 
         public static AnonymizationFhirPathRule CreateAnonymizationFhirPathRule(Dictionary<string, object> config)
         {
             if (config == null)
             {
-                throw new ArgumentNullException("config");
+                throw new ArgumentNullException(nameof(config));
```
[the remark this answers is not in the document]
Comment (BoyaWu10) [reply FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerConfigurations/AnonymizationFhirPathRule.cs:27]: Thanks, fixed in the update.

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs [hunk 6]
```diff
@@ -25,49 +25,31 @@ public static ElementNode Anonymize(this ElementNode node, AnonymizationFhirPath
             return node;
         }
 
-        // Remove null children of current node, and return true => current node is null
-        public static bool RemoveNullChildren(this ElementNode node)
+        public static void RemoveNullChildren(this ElementNode node)
         {
-            if (node == null)
```
[the remark this answers is not in the document]
Comment (BoyaWu10) [reply FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs:31]: If the node is null, the function is expected to return directly. It is fixed in the update. Test cases are added as well.

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs (hunk 3 above)
Comment (BoyaWu10) [reply FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs:152]: Sure. A private function is added to hold the shared logics.

**Review by moria97 — APPROVED** [review]
🚀

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs [hunk 7]
```diff
@@ -25,49 +25,36 @@ public static ElementNode Anonymize(this ElementNode node, AnonymizationFhirPath
...
+        public static void RemoveNullChildren(this ElementNode node)
         {
             if (node == null)
             {
-                return true;
+                return;
             }
 
-            var children = node.Children().CastElementNodes().ToList();
+            var children = node.Children().ToList();
```
Comment (qiwjin) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs:35]: can we remove ToList()

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs [hunk 8]
```diff
@@ -25,49 +25,36 @@ public static ElementNode Anonymize(this ElementNode node, AnonymizationFhirPath
             return node;
         }
 
-        // Remove null children of current node, and return true => current node is null
-        public static bool RemoveNullChildren(this ElementNode node)
+        public static void RemoveNullChildren(this ElementNode node)
```
Comment (qiwjin) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs:28]: the function name is confused, what's the definition of null children?

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs [hunk 9]
```diff
@@ -25,49 +25,36 @@ public static ElementNode Anonymize(this ElementNode node, AnonymizationFhirPath
...
-            {
-                return true;
-            }
-            else
-            {
-                return false;
-            }
         }
 
         public static void AddSecurityTag(this ElementNode node, ProcessResult result)
```
Comment (qiwjin) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs:50]: I feel this function is not belong to this class @sowu880

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeVisitorExtensions.cs [hunk 10]
```diff
@@ -12,9 +11,9 @@ public static void Accept(this ElementNode node, AbstractElementNodeVisitor visi
 
             if (shouldVisitChild)
             {
-                foreach (var child in node.Children().CastElementNodes())
+                foreach (var child in node.Children())
```
Comment (qiwjin) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeVisitorExtensions.cs:14]: does it still follow Visitor Pattern?

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs (hunk 9 above)
Comment (qiwjin) [reply FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs:50]: I feel ProcessResult doesn't belong to this node

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs [hunk 11]
```diff
@@ -67,6 +68,9 @@ public override void EndVisit(ElementNode node)
 
         private ProcessResult ProcessResourceNode(ElementNode node)
         {
+            // Initialize cache for every resource node
```
Comment (qiwjin) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs:71]: consider to use ‘IAnonymizerProcessor’ interface

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs [hunk 12]
```diff
@@ -85,34 +89,109 @@ private ProcessResult ProcessResourceNode(ElementNode node)
...
-                    /** 
-                     * Special case handling:
-                     * Senario: FHIR path only contains resourceType: Patient, Resource. 
-                     * Sample AnonymizationFhirPathRule: { "path": "Patient", "method": "keep" }
-                     * 
-                     * Current FHIR path lib do not support navigate such ResourceType FHIR path from resource in bundle.
-                     * Example: navigate with FHIR path "Patient" from "Bundle.entry[0].resource[0]" is not support
-                     */
-                    matchNodes = new List<ElementNode>() { node };
+                    resultOnRule.Update(ProcessNodeRecursive((ElementNode)matchNode.ToElement(), _processors[method], context, rule.RuleSettings));
```
Comment (sowu880) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs:95]: Is it possible the matchNode be a scopedNode?

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Stu3.FunctionalTests/VersionSpecificTests.cs [hunk 13]
```diff
@@ -55,9 +56,12 @@ public void GivenAR4OnlyResource_WhenAnonymizing_ExceptionShouldBeThrown(string
         {
             AnonymizerEngine engine = new AnonymizerEngine("stu3-configuration-sample.json");
             string testContent = File.ReadAllText(ResourceTestsFile(testFile));
-            var ex = Assert.Throws<FormatException>(() => engine.AnonymizeJson(testContent));
+            var exception = Assert.Throws<InvalidInputException>(() => engine.AnonymizeJson(testContent));
             var expectedError = "type (at Cannot locate type information for type '" + ResourceName + "')";
-            Assert.Equal(expectedError, ex.Message.ToString());
+            
+            Assert.StartsWith("The input FHIR resource JSON in invalid: ", exception.Message);
```
Comment (sowu880) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Stu3.FunctionalTests/VersionSpecificTests.cs:62]: in => is

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs [hunk 14]
```diff
@@ -138,5 +136,29 @@ private void InitializeProcessors(AnonymizerConfigurationManager configurationMa
...
+
+        private ITypedElement ParseJsonToTypedElement(string json)
+        {
+            try
+            {
+                return FhirJsonNode.Parse(json).ToTypedElement(_provider);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidInputException($"The input FHIR resource JSON in invalid: {json}", ex);
```
Comment (sowu880) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs:148]: in => is

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs [hunk 15]
```diff
@@ -20,54 +20,41 @@ public static ElementNode Anonymize(this ElementNode node, AnonymizationFhirPath
...
-                if (RemoveNullChildren(child))
+                var elementNodeChild = (ElementNode)child;
+
+                // Remove empty nodes recursively
+                RemoveEmptyNodes(elementNodeChild);
+
+                if (IsEmptyNode(elementNodeChild))
                 {
-                    node.Remove(child);
+                    node.Remove(elementNodeChild);
```
Comment (sowu880) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs:45]: seems elementNodeChild could not be null here.

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs (hunk 14 above)
Comment (BoyaWu10) [reply FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs:148]: Thanks, fixed in the update.

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs (hunk 12 above)
Comment (BoyaWu10) [reply FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs:95]: Yes, it is converted to ElementNode with `ToElement()`.

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Stu3.FunctionalTests/VersionSpecificTests.cs (hunk 13 above)
Comment (BoyaWu10) [reply FHIR/src/Microsoft.Health.Fhir.Anonymizer.Stu3.FunctionalTests/VersionSpecificTests.cs:62]: Thanks, fixed in the update.

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs (hunk 9 above)
Comment (BoyaWu10) [reply FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs:50]: By offline discussion, it is moved to ResourceProcessor.

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs (hunk 11 above)
Comment (BoyaWu10) [reply FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs:71]: Sure. ResourceProcessor is added in the update.

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs (hunk 15 above)
Comment (BoyaWu10) [reply FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs:45]: Yes, `elementNodeChild` is a child of an ElementNode. The children of ElementNode should never been null. In the library, the children names are accessed directly without checking null: https://github.com/FirelyTeam/firely-net-common/blob/fbc8608775ee6c687031214b4b4d64cb47bfa135/src/Hl7.Fhir.ElementModel/DomNode.cs#L22. Besides, in the original `RemoveNullChildren`, the logic is the same.

**Review by qiwjin — APPROVED** [review]
much clearer

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs [hunk 16]
```diff
@@ -122,5 +109,16 @@ public static void AddSecurityTag(this ElementNode node, ProcessResult result)
                 node.Replace(s_provider, metaNode, newMetaNode);
             }
         }
+
+        private static bool IsEmptyNode(ITypedElement node)
+        {
+            // A node is considered empty when: 1) it is null; 2) it has no children and its value is null.
```
Comment (sowu880) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Extensions/ElementNodeOperationExtensions.cs:45]: if child will never be null, the node here will never be null?

File: FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs [hunk 17]
```diff
@@ -71,7 +68,7 @@ public ITypedElement AnonymizeElement(ITypedElement element, AnonymizerSettings
             EnsureArg.IsNotNull(element, nameof(element));
             try
             {
-                ElementNode resourceNode = ElementNode.FromElement(element);
+                ElementNode resourceNode = ParseTypedElementToElementNode(element);
```
Comment (sowu880) [inline FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/AnonymizerEngine.cs:71]: settings can be removed from parameters in this function.

**Review by sowu880 — APPROVED** [review]

**Pull request merged**